common/src/main/java/com/microsoft/identity/common/internal/cache/HelloCache.java

```diff
@@ -0,0 +1,152 @@
+//  Copyright (c) Microsoft Corporation.
+//  All rights reserved.
+//
+//  This code is licensed under the MIT License.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files(the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions :
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+package com.microsoft.identity.common.internal.cache;
+
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.os.Build;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+
+import com.microsoft.identity.common.internal.logging.Logger;
+
+/**
+ * Persisted cache for the IPC hello() protocol.
+ * Use client's protocol version and the targeted app's package name and app version as a key
+ * to cache the negotiated protocol version.
+ * <p>
+ * This means a new hello() call will ONLY be triggered only when.
+ * 1. IPC operation is invoked for the very first time.
+ * 2. Client bumps up protocol version.
+ * 3. The targeted app is updated, uninstalled, reinstalled.
+ * <p>
+ * NOTE: the downside of this is that if you're debugging protocol change on the targeted app side,
+ * you'll either have to keep bumping the test app version or clear the cache.
+ * I've provided clearCache() for this purpose.
```

Comment:
Or you could pass a flag to disable/enable the cache for Hello....I would actually like to see that...doesn't necessarily have to be in this PR.

Reply:
Might as well do it otherwise it'll sit in my backlog for eternity :D